src/uki.rs

```diff
@@ -0,0 +1,183 @@
+    machine: U16,
+    number_of_sections: U16,
+    time_date_stamp: U32,
+    pointer_to_symbol_table: U32,
+    number_of_symbols: U32,
+    size_of_optional_header: U16,
+    characteristics: U16,
+}
+
+#[derive(Debug, FromBytes, Immutable, KnownLayout)]
+#[repr(C)]
+struct PeHeader {
+    pe: [u8; 4], // P E \0 \0
+    coff_file_header: CoffFileHeader,
+}
+
+#[derive(Debug, FromBytes, Immutable, KnownLayout)]
+#[repr(C)]
+struct SectionHeader {
+    name: [u8; 8],
+    virtual_size: U32,
+    virtual_address: U32,
+    size_of_raw_data: U32,
+    pointer_to_raw_data: U32,
+    pointer_to_relocations: U32,
+    pointer_to_line_numbers: U32,
+    number_of_relocations: U16,
+    number_of_line_numbers: U16,
+    characteristics: U32,
+}
+
+#[derive(FromBytes, Immutable, KnownLayout)]
+#[repr(C)]
+struct Array<T>([T]);
+
+#[derive(Debug, Error)]
+pub enum UkiError {
+    #[error("UKI is not valid EFI executable")]
+    PortableExecutableError,
+    #[error("UKI doesn't contain a '.osrel' section")]
+    MissingOsrelSection,
+    #[error("os-release is not UTF-8")]
+    UnicodeError(#[from] Utf8Error),
+}
+
+fn get_pe_section<'a>(image: &'a [u8], section_name: &[u8; 8]) -> Result<&'a [u8], UkiError> {
+    let Ok((dos_stub, _rest)) = DosStub::ref_from_prefix(image) else {
+        return Err(UkiError::PortableExecutableError);
+    };
+
+    let pe_offset = dos_stub.pe_offset.get() as usize;
+    let Ok((pe_header, rest)) = PeHeader::ref_from_prefix(&image[pe_offset..]) else {
+        return Err(UkiError::PortableExecutableError);
+    };
+
+    if &pe_header.pe != b"PE\0\0" {
+        return Err(UkiError::PortableExecutableError);
+    }
+
+    let sections_start = pe_header.coff_file_header.size_of_optional_header.get() as usize;
+    let n_sections = pe_header.coff_file_header.number_of_sections.get();
+    let Ok((sections, _rest)) = Array::<SectionHeader>::ref_from_prefix_with_elems(
+        &rest[sections_start..],
+        n_sections as usize,
+    ) else {
+        return Err(UkiError::PortableExecutableError);
+    };
+
+    for section in sections.0.iter() {
+        if &section.name == section_name {
+            let start = section.pointer_to_raw_data.get() as usize;
+            let len = section.virtual_size.get() as usize;
+            let end = start + len;
+
+            return Ok(&image[start..end]);
+        }
+    }
+
+    Err(UkiError::MissingOsrelSection)
+}
+
+fn get_os_release(image: &[u8]) -> Result<HashMap<&str, &str>, UkiError> {
+    // https://www.freedesktop.org/software/systemd/man/latest/os-release.html
+    let os_release = std::str::from_utf8(get_pe_section(image, b".osrel\0\0")?)?;
+    // We're intentionally dumb about the parsing and this may snap up comments and stuff, but we
+    // won't ever do lookups for keys starting with '#'...
+    Ok(HashMap::from_iter(
+        os_release.lines().filter_map(|line| line.split_once('=')),
+    ))
+}
+
+fn dequote(value: &str) -> Option<String> {
```

Comment:
This function seems very unit-testable

Also, there's the https://docs.rs/shlex/latest/shlex/ crate

(And for that matter https://github.com/pop-os/os-release but that thing is...really old and not maintained)

Reply:
You know me: I hate taking dependencies :)

I think the subset of shell dequoting that we need to understand here is simple enough to just do it ourselves...  I agree that some tests would be good though.